File: solutions/cpp/bank-account/2/bank_account.cpp
#include "bank_account.h"
#include <exception>
#include <optional>
#include <stdexcept>

namespace Bankaccount {
void Bankaccount::open() {
  if (saving)
    throw std::runtime_error("");
  saving = 0;
}
Bankaccount::Bankaccount() {}

int Bankaccount::balance() {
  if (!saving) {
    throw std::runtime_error("");
  }
  return saving.value();
}

void Bankaccount::deposit(int d) {
  if (!saving || d < 0)
    throw std::runtime_error("");
  calc(d);
}

void Bankaccount::withdraw(int w) {
  if (!saving || saving.value() < w || w < 0)
    throw std::runtime_error("");
  calc(-w);
}
void Bankaccount::close() {
  if (!saving)
    throw std::runtime_error("");
  saving.reset();
}

void Bankaccount::calc(int d) {
  lock.lock();
  saving = saving.value() + d;
  lock.unlock();
}

} // namespace Bankaccount

File: solutions/go/bank-account/2/bank_account.go
package account

import "sync"

// Define the Account type here.
const (
	open  = true
	close = false
)

type Account struct {
	m       sync.Mutex
	status  bool
	balance int64
}

func Open(amount int64) *Account {
	if amount < 0 {
		return nil
	}
	test2 := new(Account)
	test2.balance = amount
	test2.status = open
	return test2
}

func (a *Account) Balance() (int64, bool) {
	if !a.status {
		return 0, false
	}
	return a.balance, open
}

func (a *Account) Deposit(amount int64) (int64, bool) {
	a.m.Lock()
	defer a.m.Unlock()

	if !a.status || a.balance+amount < 0 {
		return 0, false
	}

	a.balance += amount
	return a.balance, open
}

func (a *Account) Close() (int64, bool) {
	if a.status {
		a.status = close
		return a.balance, true
	} else {
		return 0, false
	}
}
